esi/src/expression.rs

```diff
@@ -0,0 +1,1206 @@
+                ))
+            }
+        }
+    }
+
+    while let Some(c) = cur.peek() {
+        match c {
+            '0'..='9' => buf.push(cur.next().unwrap()),
+            _ => break,
+        }
+    }
+    let Ok(num) = buf.into_iter().collect::<String>().parse() else {
+        return Err(ExecutionError::ExpressionError(
+            "invalid number".to_string(),
+        ));
+    };
+    Ok(Token::Integer(num))
+}
+
+fn get_bareword(cur: &mut Peekable<Chars>) -> Token {
+    let mut buf = Vec::new();
+    buf.push(cur.next().unwrap());
+
+    while let Some(c) = cur.peek() {
+        match c {
+            'a'..='z' | 'A'..='Z' | '0'..='9' | '_' => buf.push(cur.next().unwrap()),
+            _ => break,
+        }
+    }
+    Token::Bareword(buf.into_iter().collect())
+}
+
+fn get_string(cur: &mut Peekable<Chars>) -> Result<Token> {
+    let mut buf = Vec::new();
+    let mut triple_tick = false;
+
+    if cur.peek() == Some(&'\'') {
+        // This is either an empty string, or the start of a triple tick string
+        cur.next();
+        if cur.peek() == Some(&'\'') {
+            // It's a triple tick string
+            triple_tick = true;
+            cur.next();
+        } else {
+            // It's an empty string, let's just return it
+            return Ok(Token::String(String::new()));
+        }
+    }
+
+    while let Some(c) = cur.next() {
+        match c {
+            '\'' => {
+                if !triple_tick {
+                    break;
+                }
+                if let Some(c2) = cur.next() {
+                    if c2 == '\'' && cur.peek() == Some(&'\'') {
+                        // End of a triple tick string
+                        cur.next();
+                        break;
+                    }
+                    // Just two ticks
+                    buf.push(c);
+                    buf.push(c2);
+                } else {
+                    // error
+                    return Err(ExecutionError::ExpressionError(
+                        "unexpected eof while parsing string".to_string(),
+                    ));
+                };
+            }
+            '\\' => {
+                if triple_tick {
+                    // no escaping inside a triple tick string
+                    buf.push(c);
+                } else {
+                    // in a normal string, we'll ignore this and buffer the
+                    // next char
+                    if let Some(escaped_c) = cur.next() {
+                        buf.push(escaped_c);
+                    } else {
+                        // error
+                        return Err(ExecutionError::ExpressionError(
+                            "unexpected eof while parsing string".to_string(),
+                        ));
+                    }
+                }
+            }
+            _ => buf.push(c),
+        }
+    }
+    Ok(Token::String(buf.into_iter().collect()))
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
The tests are a large part of this file, I wonder if it's worth extracting them into the `tests` directory.

Reply:
Not a blocker

Reply:
Yeah i was wondering about this. I kind of like having most of them in the file because they use functions that aren't public. But it is a lot. Not super sure.